anchor/eth/src/event_processor.rs

```diff
@@ -32,72 +39,83 @@ pub enum Mode {
+        Self {
+            db,
+            mode,
+            rpc_client,
+        }
     }
 
     /// Process a new set of logs
     #[instrument(skip(self, logs), fields(logs_count = logs.len()))]
-    pub fn process_logs(&self, logs: Vec<Log>, live: bool) {
+    pub async fn process_logs(&self, logs: Vec<Log>, live: bool) {
         info!(logs_count = logs.len(), "Starting log processing");
         let timer = metrics::start_timer(&metrics::EXECUTION_LOG_PROCESSING_TIME);
 
         for (index, log) in logs.iter().enumerate() {
             trace!(log_index = index, topic = ?log.topic0(), "Processing individual log");
 
-            // extract the topic0 to retrieve log handler
-            let topic0 = log.topic0().expect("Log should always have a topic0");
-            let handler = self
-                .handlers
-                .get(topic0)
-                .expect("Handler should always exist");
+            // Extract the topic0 to identify the event type
+            let topic0 = match log.topic0() {
+                Some(topic) => topic,
+                None => {
+                    warn!("Log missing topic0, skipping");
+                    continue;
+                }
+            };
+
+            // Use a match statement instead of a HashMap lookup
+            let result = match *topic0 {
+                hash if hash == SSVContract::OperatorAdded::SIGNATURE_HASH => {
+                    self.process_operator_added(log)
+                }
+
+                hash if hash == SSVContract::OperatorRemoved::SIGNATURE_HASH => {
+                    self.process_operator_removed(log)
+                }
+
+                hash if hash == SSVContract::ValidatorAdded::SIGNATURE_HASH => {
+                    self.process_validator_added(log)
+                }
+
+                hash if hash == SSVContract::ValidatorRemoved::SIGNATURE_HASH => {
+                    self.process_validator_removed(log)
+                }
+
+                hash if hash == SSVContract::ClusterLiquidated::SIGNATURE_HASH => {
+                    self.process_cluster_liquidated(log)
+                }
+
+                hash if hash == SSVContract::ClusterReactivated::SIGNATURE_HASH => {
+                    self.process_cluster_reactivated(log)
+                }
+
+                hash if hash == SSVContract::FeeRecipientAddressUpdated::SIGNATURE_HASH => {
+                    self.process_fee_recipient_updated(log)
+                }
+
+                hash if hash == SSVContract::ValidatorExited::SIGNATURE_HASH => {
+                    self.process_validator_exited(log).await
+                }
+                _ => {
+                    debug!(?topic0, "Unknown event signature, skipping");
+                    continue;
+                }
+            };
 
-            if let Err(e) = handler(self, log) {
+            // Handle any errors from the event processing
+            if let Err(e) = result {
```

Comment:
@Zacholme7 @dknopik That's what I mentioned some time ago, we are logging twice.

Reply:
I created https://github.com/sigp/anchor/issues/317